### PR TITLE
[PiranhaJava] Remove enum constants with field matching flag value

### PR DESCRIPTION
**Pull request by tnterdan**
Fixes https://github.com/uber/piranha/issues/141

Adds an optional top-level field to the config called `enumProperties`. What this field does, is if the user specifies an enum class name, Piranha will remove enum constants that have a constructor with an argument that matches that `FlagName` value, along with their usages.

### CODE REVIEW

**Review by mkr-plse — COMMENTED** [review]
Thanks for the PR @tnterdan. Added a few comments. Let us further iterate on `isMatchingEnumFieldValue` implementation as it is invoked for all matching FF API.

File: java/piranha/src/test/java/com/uber/piranha/PiranhaTestingHelpers.java
Comment (tnterdan) [reply java/piranha/src/test/java/com/uber/piranha/PiranhaTestingHelpers.java:35]: Good point. Will add a TODO, as this should probably be a separate fix since I believe Piranha already doesn't handle removal of the last enum constant with the semicolon properly (it only checks for enums ending with a `,` rather than `,` or `;`)

Comment (tnterdan) [reply java/piranha/src/test/java/com/uber/piranha/PiranhaTestingHelpers.java]: Created a ticket https://github.com/uber/piranha/issues/143

File: java/piranha/src/main/java/com/uber/piranha/config/PiranhaMethodRecord.java
Comment (lazaroclapp) [inline java/piranha/src/main/java/com/uber/piranha/config/PiranhaMethodRecord.java:90]: Again, feel free to just inline this, rather than making a new method just for passing `ARGUMENT_INDEX_KEY` :)

File: java/piranha/src/main/java/com/uber/piranha/config/PiranhaRecord.java
Comment (lazaroclapp) [reply java/piranha/src/main/java/com/uber/piranha/config/PiranhaRecord.java:1]: Agreed! (Also, copyright header, please :) )

File: java/piranha/src/test/java/com/uber/piranha/ConfigurationTest.java
Comment (lazaroclapp) [reply java/piranha/src/test/java/com/uber/piranha/ConfigurationTest.java:1638]: I still don't quite understand this method or its comment. What is this testing _or_ setting up for other tests?

File: java/piranha/src/test/java/com/uber/piranha/ConfigurationTest.java
Comment (lazaroclapp) [inline java/piranha/src/test/java/com/uber/piranha/ConfigurationTest.java:1655]: One line description here? (I like the link, but it should be possible to get some idea what this is about without following the link. Also, if we ever move away from GitHub for issue tracking or relocate to another repo/org, some documentation should remain in the code itself)

Comment (lazaroclapp) [reply java/piranha/src/test/java/com/uber/piranha/ConfigurationTest.java:1655]: Actually, looking at the paragraph below, this is fine, I think... though maybe I would still do for these tests:

```
Testing enumProperties configuration.

<p>[the test specific paragraph you have written for each test]

<p> See also: [https://github.com/uber/piranha/issues/141]
```

File: java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java
Comment (lazaroclapp) [inline java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java:506]: Again, I'd move the issue links to the end, start with a line on what is being tested (which can be the existing next line)

File: java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java
Comment (lazaroclapp) [inline java/piranha/src/test/java/com/uber/piranha/CorePiranhaTest.java:514]: We "recently" put some overdue work to partition `CorePiranhaTest` into multiple test classes, which is both better for readability and allows running our tests in parallel with gradle.

Could we get all tests related to enum constant support into their own test suite class? (I know that's a small amount of boilerplate, but we really don't want to end up with 2-3k  `CorePiranhaTest` again 😅

**Review by lazaroclapp — CHANGES_REQUESTED** [review]
Overall this is all really cool work! A few comments/changes requested, though :)

Also, we land PRs by squashing, so if you could change the description to something suitable to be the commit message for this PR, then that would be appreciated.

Again, thanks so much for contributing!

File: java/piranha/src/main/java/com/uber/piranha/config/PiranhaEnumRecord.java
Comment (tnterdan) [reply java/piranha/src/main/java/com/uber/piranha/config/PiranhaEnumRecord.java:14]: Added!

File: java/piranha/src/test/java/com/uber/piranha/ConfigurationTest.java
Comment (tnterdan) [reply java/piranha/src/test/java/com/uber/piranha/ConfigurationTest.java:1638]: Good point, the `Test` in the name isn't helping here... I'll think of something clearer. Or it may also be cleaner to keep some duplication in tests for readability's sake.

**Comment by tnterdan** [comment]
Just want to say thanks an absolute ton for the review comments so far @mkr-plse and @lazaroclapp 👏 I know how time consuming code reviews can be, so I super appreciate your attention to detail! 🌟

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java
Comment (tnterdan) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:431]: Hmm so after thinking about this, the issue is if we remove the constant `STALE_FLAG`, it doesn't matter what method call we find, we need to remove all usages because we removed the enum constant itself.

There is a potential use case where the constant stays, and we remove ONLY based on method call. As of right now, that doesn't seem like a useful use case to me. (I could be wrong though!)

Will add a comment to clarify at the very least.

Comment (lazaroclapp) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:431]: Haven't done a full second pass yet, but that's a very valid point. Since the methods are instance methods, we also can't associate them just with the enum, so, removing the enum constant requires removing any calls to methods of that enum constant. Now, the question is, how do we simplify code around that call. Consider the following:

```
public enum ButtonColorFlag {
    // ... a hundred+ other flags
    BLUE_BUTTON_FLAG("button.is.blue", "#0000FF"),
    RED_BUTTON_FLAG("button.is.red", "#FF0000");

    private final String key;
    private final String value;

    public String getKey() {
        return key;
    }

    public String getValue() {
        return value;
    }

    FeatureFlags(final String key, final String value) {
        this.key = key;
        this.value = value;
    }
}
```

Now, presumably code calling `BLUE_BUTTON_FLAG.getKey()` will only happen inside whatever checks for whether the feature is toggled or not (`featureFlagClient.isFeatureEnabled` in #141), but `getValue()` could be called directly, with the value meaning something and, more importantly being assigned or passed to some other field/call:

```
this.button = new Button(color=BLUE_BUTTON_FLAG.getValue())
```

So, what is Piranha to do then?

a) Abort? (e.g. we just don't allow flag enums to carry values others than those needed to check the flag name/identifier)
b) Remove the line? Possibly leaving `this.button` unassigned and breaking code? (In our internal deployment, we don't require Piranha to always produce building code, as some refactorings will require manual follow up no matter what, but we do aim for it)
c) Try to propagate `value`? (Easy-ish in the example above, but impossible in the general case where the enum contains methods that aren't statically evaluatable)

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java
Comment (mkr-plse) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java]: Here is another proposal. I believe the enum declarations will be processed by ErrorProne before the invocations are processed. Is it possible to infer the enum flag declarations and methods when the enum class is parsed, store the results and reuse it in `getXPAPI`? 

This will be a problem if the enum declarations are in a different compilation unit (or in a library jar) and are not part of the current compilation. To overcome this, we can have an option to check for the stored results and if unavailable, perform the computation that you have currently, and this option can be turned off by default. 

I think this inference and use will suffice in a majority of the cases.

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java
Comment (tnterdan) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:468]: So far I tried to query members of the symbol for the enum class and was able to get the members. However, it seems that the symbol table doesn't hold the actual `"stale.flag"` String value from the constructor.  At least I don't see it in whatever implementation of the JVM I'm on (a version of AdoptOpenJDK 1.8) in the `com.sun.tools.javac.code` package. But I assume the String value has to be accessible outside of source (such as in the bytecode), so I'll keep investigating on how to get it for other targets/jars.

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java
Comment (tnterdan) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java]: Isn't this where two passes are needed? Because we can't rely on the enum class being parsed before its usages?

Although I am happy to cache results for multiple reasons.

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java
Comment (tnterdan) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:422]: Using a switch statement has some scoping issues where I couldn't reuse declarations of `memberTree` within case blocks and would have to either rename one or both `memberTree` variables, or declare it once in the outer scope. So I'm not sure if that's preferred to using a switch or not.

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java
Comment (tnterdan) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java]: Agreed and added caching :)

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java
Comment (tnterdan) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:487]: Kind of thinking about this later, we need the source to be able to remove the enum constant in the first place? Does Piranha already have tests that cover the case where we have an enum in another target/jar, and therefore possibly no access to source code, but we clean up the enum's usages anyway while preserving the enum constant?

I'm really not sure how to get access to the constructor argument values from the symbol table. I don't see it represented anywhere in memory.

To avoid having this block the rest of the PR, may be another ticket to do later, and I can add a TODO if that's OK? At least it will work for projects that have their flags self-contained. I think regardless it would be good to add tests for feature flags across multi-module projects like your example.

**Comment by tnterdan** [comment]
OK I believe I have addressed or responded to everything brought up so far 👍 Thank you @mkr-plse and @lazaroclapp, let me know if I missed anything (or if you have any ideas on how to get the constructor argument values out of the symbol table, I tried!) Looking forward to your next pass.

File: java/README.md
Comment (mkr-plse) [inline java/README.md:279]: Given Piranha removes the method call, once it is configured, I am not sure whether this merits a discussion in the README. This is expected behavior even for non-enum usages where the method cleanup will not happen if the method is not configured. Can we get rid of this discussion?

**Review by mkr-plse — COMMENTED** [review]
Overall, looks good to me except for the update to README.

**Comment by mkr-plse** [comment]
> Overall, looks good to me except for the update to README.

Also, `java/piranha/src/test/java/com/uber/piranha/PiranhaTestingHelpers.java` needs to be updated to handle GJF (the build seems to have failed due to this).

File: java/README.md
Comment (tnterdan) [reply java/README.md]: Yeah wasn't sure if it was helpful, removed. I purposefully put it in its own commit though so it was easy to remove 😂

**Comment by tnterdan** [comment]
OK formatting should be fixed, ran `./gradlew verifyGoogleJavaFormat` locally to verify 👍 Thanks for reviews @mkr-plse and @lazaroclapp :)

**Review by mkr-plse — APPROVED** [review]
LGTM. Thanks for the PR again.

**Comment by tnterdan** [comment]
Just waiting on a re-review form @lazaroclapp :)

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java
Comment (lazaroclapp) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:438]: Good point. I actually just made a quick test case for it. You are correct, (b) is the current behavior, and this change preserves it, which makes sense to me. Wonder if we could in the future add some logic to Piranha to replace statically evaluatable methods of enum constant with their constant values but, in the limit, this is impossible to handle for all cases of enum methods.

File: java/piranha/src/test/java/com/uber/piranha/EnumConstantTest.java
Comment (lazaroclapp) [inline java/piranha/src/test/java/com/uber/piranha/EnumConstantTest.java:33]: 👍  on having this be its own test class!

**Review by lazaroclapp — APPROVED** [review]
Hey @tnterdan , sorry for the delay on my review...

Below are a few comments and one or two very minor nits, but overall, I think this looks ready to land!

Might need to be rebased, though :)

**Comment by tnterdan** [comment]
Rebased and squashed already reviewed changes. Thanks for the re-review @lazaroclapp! No worries about it taking time, we all have other priorities to attend to :)

Yeah I agree it would be pretty nice to find the enum constructor without the source, but yeah, I think it's worth following up on that with some better multi-target testing and digging into the symbol table API a bit more. At least it's behind a config option so existing users won't be in for any surprises.

I like to avoid putting too many changes into a single PR anyway 😅 I'm sure this was already a doozy to review.

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java
Comment (lazaroclapp) [inline java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:505]: 👍  (There is a chance that we discover that in practice erroring out here is too strict and we should print a warning, but I definitely prefer to start with the most "fail-fast" possible solution until we know for sure how often this case shows up or if there is any better way to handle it! :) )

File: java/piranha/src/main/java/com/uber/piranha/PiranhaRuntimeException.java
Comment (lazaroclapp) [inline java/piranha/src/main/java/com/uber/piranha/PiranhaRuntimeException.java:1]: As silly as that might be for this tiny exception class... can we get a copyright header here?

**Review by lazaroclapp — CHANGES_REQUESTED** [review]
One small fix requested regarding the copyright header for the new exception file. Other than that, this is ready to land.

File: java/piranha/src/main/java/com/uber/piranha/PiranhaRuntimeException.java
Comment (tnterdan) [reply java/piranha/src/main/java/com/uber/piranha/PiranhaRuntimeException.java:14]: Oops! This is what I get for copying and pasting from `PiranhaConfigurationException` 😂

File: java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java
Comment (tnterdan) [reply java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java:505]: I'm generally in favor of failing fast 😁 Hiding errors tends to cause problems later...

**Comment by tnterdan** [comment]
Thanks so much for the reviews @lazaroclapp and @mkr-plse!

File: java/piranha/src/main/java/com/uber/piranha/PiranhaRuntimeException.java
Comment (lazaroclapp) [reply java/piranha/src/main/java/com/uber/piranha/PiranhaRuntimeException.java:14]: 😅  yup, that one is on us... I need to go over the code and fix those java files that are still missing the header.

**Review by lazaroclapp — APPROVED** [review]
Ready to merge if CI passes.

**Comment by tnterdan** [comment]
I'm seeing this message:
> Only those with write access to this repository can merge pull requests.

@lazaroclapp or @mkr-plse could one of you merge on my behalf? Thank you! 🙏

**Comment by lazaroclapp** [comment]
> I'm seeing this message:
> 
> > Only those with write access to this repository can merge pull requests.
> 
> @lazaroclapp or @mkr-plse could one of you merge on my behalf? Thank you! 🙏

It is done! 🎉 

Thanks again for the PR and for being patient with our reviews and comments!

**Comment by mkr-plse** [comment]
Thanks @tnterdan and @lazaroclapp. Glad to see this land :)

**Pull request merged**
